### PR TITLE
T/fix close connection on reject

### PR DESCRIPTION
**Pull request by tspopp**

### CODE REVIEW

File: rsocket/src/transport/socket.rs [hunk 1]
```diff
@@ -64,15 +64,26 @@ impl DuplexSocket {
             abort_handles: Arc::new(DashMap::new()),
         };
 
-        let cloned_socket = socket.clone();
```
Comment (tspopp) [inline rsocket/src/transport/socket.rs:67]: Dropping `tx` is used to signal that this `DuplexSocket` instance is no longer producing any Frames.

If we create a copy of the duplex socket, we are not able to ever drop the `tx` completely, because this clone will always hold a valid reference to `tx`. Therefore, the `loop_canceller` function has been adapted, not requiring a clone of `DuplexSocket` anymore.

File: rsocket/src/transport/socket.rs [hunk 2]
```diff
@@ -64,15 +64,26 @@ impl DuplexSocket {
...
+    fn tx(&self) -> Result<&mpsc::UnboundedSender<Frame>> {
+        if self.tx.is_none() {
+            return Err(anyhow!("connection has been closed!"));
+        }
+        Ok(self.tx.as_ref().unwrap())
+    }
+
+    pub fn close(&mut self, msg: &str) {
+        warn!("closed channel: {msg}");
+        self.tx.take();
```
Comment (tspopp) [inline rsocket/src/transport/socket.rs:84]: Convenient method to `drop` the `tx`. Dropping `tx` will implicitly emit a `None` value, which stops our Network reader and writer tasks.

File: rsocket/src/transport/socket.rs [hunk 3]
```diff
@@ -121,6 +133,12 @@ impl DuplexSocket {
         let sid = msg.get_stream_id();
         let flag = msg.get_flag();
         debug_frame(false, &msg);
+
+        if self.tx().is_err() {
```
Comment (tspopp) [inline rsocket/src/transport/socket.rs:137]: In case we receive a new `Frame` from the network and the Duplex Socket has been already closed, we ignore the frame and print a warning.

Any (inlined) methods called from `process_once` are able to `unwrap()`the `tx()` without further error checking.

File: rsocket/src/transport/socket.rs [hunk 4]
```diff
@@ -379,7 +399,7 @@ impl DuplexSocket {
     #[inline]
     fn send_cancel_frame(&self, sid: u32) {
         let cancel_frame = frame::Cancel::builder(sid, Frame::FLAG_COMPLETE).build();
-        if let Err(e) = self.tx.send(cancel_frame) {
+        if let Err(e) = self.tx().unwrap().send(cancel_frame) {
```
Comment (tspopp) [inline rsocket/src/transport/socket.rs:402]: These `unwrap()` calls are ok, since `process_once` has already checked that `tx` is still available.

File: rustfmt.toml [hunk 5]
```diff
@@ -1,2 +1,2 @@
-edition = 2021"
+edition = "2021"
```
Comment (tspopp) [inline rustfmt.toml:1]: `fmt` was not working due to that TOML typo

File: rsocket/src/core/server.rs [hunk 6]
```diff
@@ -124,26 +127,43 @@ where
                     break;
                 }
             }
+
+            // drop reader task in case duplex socket is no longer producing frames
+            reader_handle.abort();
```
Comment (tspopp) [inline rsocket/src/core/server.rs:132]: In case `Duplex Socket` is no longer emitting `Frames` to the network, the writer task is closed. For releasing the connection completely, we must close our reader task as well:

This handle aborts the reader task and finally drops the `connection` which finally drops the TCP connection and informs our peer.

File: rsocket/src/transport/socket.rs (hunk 2 above)
Comment (tspopp) [reply rsocket/src/transport/socket.rs:84]: I need to change this one more time. There might be more clones of `Duplex Socket` in the wild. Calling `close` will not necessary lead in `tx` being dropped and therefore the warning is misleading. `tx` is dropped if all instances of `Duplex Socket` closed their `tx`. This applies to the `server` implementation, but not to the `client` implementation where `Duplex Socket` is cloned in multiple occassions.

The `Clonable` attribute of `Duplex Socket` feels weird. Maybe we can go for an Arc<DuplexSocket>, enfording a  only a single instance of DuplexSocket 

- [ ] lock free, non cloneable Duplex Socket ?

File: .gitignore [hunk 7]
```diff
@@ -3,3 +3,4 @@ target/
 Cargo.lock
 .vscode/
 /flamegraph.svg
+.idea
```
Comment (adoerr) [inline .gitignore:6]: I tend to use `**/.idea` in order to cover sub-directories as well.

File: Cargo.toml [hunk 8]
```diff
@@ -9,3 +9,7 @@ members = [
   "examples",
   "rsocket-test",
 ]
+
+[patch.crates-io]
+"rsocket_rust" = { path = "../rsocket-rust/rsocket" }
+"rsocket_rust_transport_tcp" = { path = "../rsocket-rust/rsocket-transport-tcp" }
```
Comment (adoerr) [inline Cargo.toml:15]: What is the idea behind patching the release crates here?

File: rsocket-test/tests/test_setup_rejection.rs [hunk 9]
```diff
@@ -0,0 +1,59 @@
+#[macro_use]
+extern crate log;
```
Comment (adoerr) [inline rsocket-test/tests/test_setup_rejection.rs:2]: Since the `log` crate macros are all `#[macro_export]` you can use them with a regular path. So `use log::*` or something will do.

File: rsocket-test/tests/test_stream_cancellation.rs [hunk 10]
```diff
@@ -59,7 +61,12 @@ mod tests {
     /// When request_stream handle is dropped: 0 subscribers
     async fn request_stream_can_be_cancelled_by_client(client: Client) {
         assert_eq!(
-            client.request_response(Payload::from("subscribers")).await.unwrap().unwrap().data_utf8(),
+            client
+                .request_response(Payload::from("subscribers"))
+                .await
+                .unwrap()
+                .unwrap()
```
Comment (adoerr) [inline rsocket-test/tests/test_stream_cancellation.rs:68]: This is a nit really but perhaps `Option::flatten()` helps to avoid all those ugly double-unwraps :thinking:

File: rustfmt.toml (hunk 5 above)
Comment (adoerr) [reply rustfmt.toml:1]: I think, it might have been myself, introducing this error :see_no_evil:

**Review by adoerr — APPROVED** [review]
Can't really say much about the general approach due to lack of familiarity with RSocket.

Other than that, LGTM

File: Cargo.toml (hunk 8 above)
Comment (tspopp) [reply Cargo.toml:15]: It seems that subprojects have dependencies on `rsocket-rust` within `crates.io`. This patch forces to build locally, without `rsocket-rust` from `crates.io`. Without this patch I wasn't able to build locally.

**TLDR: This patch should be reverted before making pushing the PR to the origin repository**
